domain/src/main/java/org/oppia/android/domain/platformparameter/syncup/PlatformParameterSyncUpWorker.kt

```diff
@@ -0,0 +1,112 @@
+  private val platformParameterService: PlatformParameterService,
+  private val oppiaLogger: OppiaLogger,
+  private val exceptionsController: ExceptionsController,
+  @BackgroundDispatcher private val backgroundDispatcher: CoroutineDispatcher
+) : CoroutineWorker(context, params) {
+
+  companion object {
+    /** Exception message when the type of values received in the network response are not valid. */
+    const val INCORRECT_TYPE_EXCEPTION_MSG =
+      "Platform parameter value has incorrect data type, ie. other than String/Int/Boolean"
+
+    /** Exception message when there are no values received in the network response. */
+    const val EMPTY_RESPONSE_EXCEPTION_MSG = "Received an empty map in the network response"
+
+    /** A Tag for the logs that are associated with PlatformParameterSyncUpWorker. */
+    const val TAG = "PlatformParameterWorker.tag"
+
+    /** Value of worker-type associated with this PlatformParameterSyncUpWorker. */
+    const val PLATFORM_PARAMETER_WORKER = "platform_parameter_worker"
+
+    /** Key for passing the worker-type as a parameter to PlatformParameterSyncUpWorker. */
+    const val WORKER_TYPE_KEY = "worker_type_key"
+  }
+
+  override suspend fun doWork(): Result {
+    return when (inputData.getString(WORKER_TYPE_KEY)) {
+      PLATFORM_PARAMETER_WORKER -> withContext(backgroundDispatcher) { refreshPlatformParameters() }
+      else -> Result.failure()
+    }
+  }
+
+  // Parses a map of platform parameter values into a [List<PlatformParameter>]. If the parameters
+  // are not of type String, Int or Boolean this function fails with an [IllegalArgumentException].
+  private fun parseNetworkResponse(response: Map<String, Any>): List<PlatformParameter> {
+    val platformParameterList: MutableList<PlatformParameter> = mutableListOf()
+    for (entry in response.entries) {
+      val platformParameter = PlatformParameter.newBuilder().setName(entry.key)
+      when (val value = entry.value) {
+        is String -> platformParameter.string = value
+        is Int -> platformParameter.integer = value
+        is Boolean -> platformParameter.boolean = value
+        else -> throw IllegalArgumentException(INCORRECT_TYPE_EXCEPTION_MSG)
+      }
+      platformParameterList.add(platformParameter.build())
+    }
+    return platformParameterList
+  }
+
+  /** Extracts platform parameters from the remote service and stores them in the cache store */
+  private fun refreshPlatformParameters(): Result {
+    return try {
+      val response = platformParameterService.getPlatformParametersByVersion(
+        applicationContext.getVersionName()
+      ).execute()
+      val responseBody = checkNotNull(response.body())
+      val platformParameterList = parseNetworkResponse(responseBody)
+      if (platformParameterList.isEmpty()) {
+        throw IllegalArgumentException(EMPTY_RESPONSE_EXCEPTION_MSG)
+      }
+      platformParameterController.updatePlatformParameterDatabase(platformParameterList)
```

Comment:
Understood. Have implemented the function in a different way now. PTAL

Reply:
I thought that tests are passing hence the functions are working fine. Added an observerForever over here, not sure about it PTAL.

Reply:
Just to let you know, after adding observeForever on the dataProvider (which was converted to LiveData after `toLiveData( )`) some of the tests in the PlatformParameterWorkerTest started failing because of this error -> Cannot invoke observeForever on a background thread.

Reply:
We shouldn't ever use LiveData in the domain layer since it's meant to be a UI-only construct (the lack of Lifecycle being a key reason).

Could we return a Deferred or DataProvider and observe/fetch from that, instead? It probably means this does need to be a suspend function, but that's presumably fine given the execution context.

Reply:
@ARJUPTA this appears to be unresolved still.

Reply:
@BenHenning I don't know how but I have changed the code to this format.
![Screenshot 2021-08-19 114619](https://user-images.githubusercontent.com/64064110/130017669-8e1579ae-fb1d-4e9e-9ddc-fab08338f66a.png)
 The code that you are reffering to is outdated one

Reply:
Yeah, not sure what happened here. The latest implementation looks solid.